### PR TITLE
Add option to snoop commands to control spawning separate process vs eval'ing in existing process

### PR DESCRIPTION
**Pull request by NHDaly**
Running snoop compilation in a separate process allows you to consistently measure complete compilation time on every run, without being affected by the state of the current process.

On the other hand, running the snoop compilation in the current process allows you to measure _incremental_ compilation performance, which can also be useful in some contexts.

This PR attempts to add an option to all snoop commands to let you choose between those options.
It also adds another function, `@snoop_all`, which runs `@snoopi`, `@snoopl` and `@snoopc` together at the same time, so that you can compare results across them.

-----------------------

Note that this branch is currently really messy, because it had a bunch of draft PRs merged into it, that have since been squash-merged into master, so we need to clean that up first as well.

This addresses this comment, for a request from @bkamins: https://github.com/timholy/SnoopCompile.jl/issues/251#issuecomment-886093390

### CODE REVIEW

**Comment by bkamins** [comment]
Thank you. Can you please ping me when this is ready to test? (or if you want me to review it now I can also try but I do not know the code base much)

File: SnoopCompileCore/src/SnoopCompileCore.jl [hunk 1]
```diff
@@ -23,4 +23,8 @@ if VERSION >= v"1.6.0-DEV.1192"  # https://github.com/JuliaLang/julia/pull/37136
     include("snoopl.jl")
 end
 
+if VERSION >= v"1.6.0-DEV.1192"  # https://github.com/JuliaLang/julia/pull/37136
+    include("snoop_all.jl")
```
Comment (timholy) [inline SnoopCompileCore/src/SnoopCompileCore.jl:27]: Can't we just include this macro into `SnoopCompileCore/src/snoopl.jl`? It's the same condition for including it that we use for`@snoopl`.

**Review by timholy — COMMENTED** [review]
LGTM, just one small comment.

A test would be nice, of course!

File: SnoopCompileCore/src/SnoopCompileCore.jl (hunk 1 above)
Comment (NHDaly) [reply SnoopCompileCore/src/SnoopCompileCore.jl:24]: Yeah, good idea. I've just moved it into the same if-block. I guess it doesn't necessarily belong in `snoopl.jl` because it will snoop all three types of compilation, which is why I put it in its own file.

Can you have another look and let me know if this looks okay? :)

**Comment by NHDaly** [comment]
So i think this is more or less ready to go, but for some reason that I can't explain, the first call to `@snoopl` is now failing with an `IOError: write: broken pipe (EPIPE)` during the `serialize()`. But only the first time, when i run it in the REPL, and then it seems to succeed consistently after that.

I have no idea why. In the last commit, 995ffd2, I tried simplifying the logic that's spawning the processes and passing the commands through, but it had no effect.

This is also why the tests are failing.

Any advice you have would be greatly appreciated!! 😬 🤞 

=================

Otherwise, I've added tests and a docstring.

There's still no standalone tests for `@snoopc` and `@snoopl` with `pspawn=false`, and it would be nice to add those.

Finally, there's still no `pspawn=` option for `@snoopi_deep` or `@snoop_all`, which we could consider adding to make them more like the other functions. But this can be done in a follow-up PR.

**Comment by NHDaly** [comment]
Tl;dr: Tests failing and i don't know why - plz help!! 😬

**Comment by NHDaly** [comment]
@bkamins - you could try testing this now if you'd like

**Comment by bkamins** [comment]
Thank you. I do not get the error you report. However I have some general minor comments:
* running this macro swamps my terminal with the output; is this intended?
* the two files you give .csv extension are not comma separated (it seems they are tab separated)
* could you please guide me what do the numbers in `*snoopc.csv` file mean (these in the beginning of each line)

Thank you!

**Comment by bkamins** [comment]
Apart from that I have analyzed the results - and all works as I would expect. Superb work - thank you!

@pdeffebach - given the results of `@snoop_all` the difference between `subset` and `select` in compilation time is unfortunately due to the additional compilation that `subset` has (the extra logic it has). This is probably unavoidable (and this was expected).

**Comment by NHDaly** [comment]
> * running this macro swamps my terminal with the output; is this intended?

I've added an example with a `;` at the end.
Probably we should change the default show for the `InferenceTimingNode` returned as the first argument (`tinf`) from `@snoopi_deep` so that it doesn't print such a big tree, as well..

> the two files you give .csv extension are not comma separated (it seems they are tab separated)
Huh, until today I thought that you could have arbitrary delimiters, but it's still called a CSV.  Should we rename them to `.tsv`?
https://superuser.com/a/577912

> could you please guide me what do the numbers in *snoopc.csv file mean (these in the beginning of each line)

Please see the updated docstring, with links to what to do with each of the results. You should be able to follow all of those separate guides for how to interpret the results. :)

**Comment by bkamins** [comment]
Thank you!

File: SnoopCompileCore/src/snoopc.jl [hunk 2]
```diff
@@ -4,33 +4,50 @@ using Serialization
...
+function snoopc(flags, filename, commands, _module=nothing; pspawn=true)
+    if pspawn
+        println("Launching new julia process to run commands...")
+    end
     code_object = """
             using Serialization
-            while !eof(stdin)
-                Core.eval(Main, deserialize(stdin))
-            end
+            Core.eval(Main, deserialize(IOBuffer(read(stdin))))
```
Comment (timholy) [inline SnoopCompileCore/src/snoopc.jl:48]: Any particular reason for this change? What if `stdin` is *huge*? Likewise elsewhere among the changes.

Is it possible that this is the change that causes the closed stream issue? It seems possible that Julia 1.0 is not being as careful as this code.

**Review by timholy — COMMENTED** [review]
Looks quite good. For consistency, let's re-export `@snoop_all` from SnoopCompile.

File: SnoopCompileCore/src/snoopc.jl (hunk 2 above)
Comment (NHDaly) [reply SnoopCompileCore/src/snoopc.jl:48]: Alas, this is not responsible for the error -- this was my attempt to fix it, but it didn't help. I can happily revert these changes, yeah.

**Comment by kimikage** [comment]
This is not a technical issue, but the name `all` seems somewhat misleading.
Perhaps this PR needs to be targeted and updated for SnoopCompile v3 or v2, but there is room for reconsideration of the name as well.

**Comment by NHDaly** [comment]
> This is not a technical issue, but the name `all` seems somewhat misleading.

How so? What would you prefer instead? The idea was that it should perform all the analyses we have: inference, llvm, etc.
I'm open to something better. Perhaps `@snoop_all_phases`?

**Comment by NHDaly** [comment]
Also, sorry I let this PR completely rot. 😢
